### PR TITLE
test: trainer init components unit tests

### LINKED ISSUE #537: [TEST] Jax Trainer Init
### What do you want to test?
Jax trainer init components

### Outline of test structure
* Unit tests
* Test components and hooks

### Definition of done
Passing checks, cover all hooks, edge cases considered

### Mandatory checklist before making a PR
* [ ] The success criteria laid down in “Definition of done” are met.
* [ ] Test code is documented - docstrings for methods and classes, static types for arguments.
* [ ] Documentation is updated - README, CONTRIBUTING, or other documentation.

### PR DESCRIPTION
**Pull request by RuanJohn**
## What?
Unit tests for `TrainerInit` component. 
## Why?
Part of ongoing efforts to up test code coverage for Mava callback redesign.
## Extra
Closes #537

### CODE REVIEW

File: tests/jax/components/training/trainer_test.py [hunk 1]
```diff
@@ -0,0 +1,680 @@
...
+def test_single_trainer_no_shared_weights_fixed_sampling(
+    mock_builder_no_shared_weights_fixed_sampling: Builder,
+) -> None:
+    """Tests on_building_init_end hook.
+
+    Single trainer for all networks, no shared network weights and
+    fixed agent network sampling.
+    """
+
+    mock_trainer = SingleTrainerInit()
```
Comment (mmorris44) [inline tests/jax/components/training/trainer_test.py:255]: This could be a fixture, see elsewhere too. But up to you. I kinda back how it is at the moment, given the number of fixtures already.

File: tests/jax/components/training/trainer_test.py [hunk 2]
```diff
@@ -0,0 +1,680 @@
...
+def test_custom_trainer_on_building_init_value_error(
+    mock_builder_no_shared_weights_fixed_sampling: Builder,
+) -> None:
+    """Tests on_building_init_end hook.
+
+    Custom trainer init. Asserts than exception is thrown when an empty dictionary
+    is passed in.
+    """
+
+    trainer_init = CustomTrainerInit(config=CustomTrainerInitConfig())
```
Comment (mmorris44) [inline tests/jax/components/training/trainer_test.py:395]: Perhaps also an assert in here for when no dictionary is provided?

**Review by mmorris44 — CHANGES_REQUESTED** [review]
Huge PR, but really well done! Only found a couple of things I might want to change.
I'm reasonably sure that these tests cover all of the major logic paths in the code, but can't assert that for sure.
Thanks for slogging through this @RuanJohn :D

File: tests/jax/components/training/trainer_test.py [hunk 3]
```diff
@@ -0,0 +1,680 @@
...
+from mava.systems.jax.builder import Builder
+from mava.systems.jax.trainer import Trainer
+
+#################
+# Mock builders
+#################
+
+
+@pytest.fixture
+def mock_builder_shared_weights_fixed_sampling() -> Builder:
```
[the remark this answers is not in the document]
Comment (RuanJohn) [reply tests/jax/components/training/trainer_test.py:71]: Agreed, I made some classes that contain most of the reused code.

File: tests/jax/components/training/trainer_test.py (hunk 1 above)
Comment (RuanJohn) [reply tests/jax/components/training/trainer_test.py]: Made them into fixtures :call_me_hand:

File: tests/jax/components/training/trainer_test.py (hunk 2 above)
Comment (RuanJohn) [reply tests/jax/components/training/trainer_test.py:430]: Added a case where nothing is passed in as well.
